framework/src/main/java/org/checkerframework/common/accumulation/AccumulationAnnotatedTypeFactory.java

```diff
@@ -0,0 +1,288 @@
+     *     argument named "value" whose type is a String array.
+     * @param t the top type in the hierarchy
+     * @param bot the bottom type in the hierarchy
+     */
+    protected AccumulationAnnotatedTypeFactory(
+            BaseTypeChecker checker,
+            Class<? extends Annotation> accumulator,
+            Class<? extends Annotation> t,
+            Class<? extends Annotation> bot) {
+        super(checker);
+
+        TOP = AnnotationBuilder.fromClass(elements, t);
+        BOTTOM = AnnotationBuilder.fromClass(elements, bot);
+        ACC = accumulator;
+
+        // Every subclass must call postInit!
+        if (this.getClass() == AccumulationAnnotatedTypeFactory.class) {
+            this.postInit();
+        }
+    }
+
+    /**
+     * Creates a new instance of the accumulator annotation that contains the elements of {@code
+     * values} in sorted order.
+     *
+     * @param values the arguments to the annotation
+     * @return an annotation mirror representing the accumulator annotation with values's arguments,
+     *     or top is {@code values} is empty
+     */
+    public AnnotationMirror createAccumulatorAnnotation(final String... values) {
+        if (values.length == 0) {
+            return TOP;
+        }
+        AnnotationBuilder builder = new AnnotationBuilder(processingEnv, ACC);
+        Arrays.sort(values);
+        builder.setValue("value", values);
+        return builder.build();
+    }
+
+    /**
+     * Utility method that returns whether the return type of the given method invocation tree has
+     * an @This annotation from the Returns Receiver Checker.
+     *
+     * @param tree the method invocation tree to check
+     * @return whether the method being invoked returns its receiver
+     */
+    public boolean returnsThis(final MethodInvocationTree tree) {
+        ReturnsReceiverAnnotatedTypeFactory rrATF =
+                getTypeFactoryOfSubchecker(ReturnsReceiverChecker.class);
+        ExecutableElement methodEle = TreeUtils.elementFromUse(tree);
+        AnnotatedTypeMirror methodAtm = rrATF.getAnnotatedType(methodEle);
+        AnnotatedTypeMirror rrType =
+                ((AnnotatedTypeMirror.AnnotatedExecutableType) methodAtm).getReturnType();
+        return rrType != null && rrType.hasAnnotation(This.class);
+    }
+
+    /**
+     * Is the given annotation an accumulator annotation?
+     *
+     * @param anm an annotation mirror
+     * @return whether the annotation mirror is an instance of this factory's accumulator annotation
+     */
+    public boolean isAccumulatorAnnotation(AnnotationMirror anm) {
+        return AnnotationUtils.areSameByClass(anm, ACC);
+    }
+
+    @Override
+    protected TreeAnnotator createTreeAnnotator() {
+        return new ListTreeAnnotator(
+                super.createTreeAnnotator(), new AccumulationTreeAnnotator(this));
+    }
+
+    /** Handles fluent APIs using the Returns Receiver Checker. */
```

Comment:
Can we add a bit more detail on what this tree annotator is actually doing?  Like the example we give in the transfer function documentation.  This can be in the class documentation or the docs of `visitMethodInvocation()`.

Reply:
I added it to the docs for `getMethodInvocation()`